### PR TITLE
Storage: support SQL session storage

### PR DESCRIPTION
**Pull request by reinkrul**
Feature https://github.com/nuts-foundation/nuts-node/pull/3065 proposed again, given the comment https://github.com/nuts-foundation/nuts-node/pull/3065#issuecomment-2117708266 in that PR.

When parties are reluctant to deploy another database server (e.g. hospitals only running a SQL Server on bare metal), it could be viable to allow session storage in SQL. But ofc., with the notice that performance may suffer.

### CODE REVIEW

File: storage/session_sql.go [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+	} else if err != nil {
+		return err
+	}
+	if time.Now().After(time.Unix(int64(record.Expires), 0)) {
+		return ErrNotFound
+	}
+	return json.Unmarshal([]byte(record.Value), target)
+}
+
+func (s sqlSessionStore) GetAndDelete(key string, target interface{}) error {
```
Comment (woutslakhorst) [inline storage/session_sql.go:127]: untested

File: storage/session_sql.go [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+	}()
+}
+
+type sqlSessionStore struct {
+	db        *gorm.DB
+	ttl       time.Duration
+	storeName string
+}
+
+func (s sqlSessionStore) Delete(key string) error {
```
Comment (woutslakhorst) [inline storage/session_sql.go:103]: Describe behaviour when running multiple nodes with slightly different clocks.....

**Review by woutslakhorst — CHANGES_REQUESTED** [review]

File: docs/pages/deployment/storage.rst [hunk 3]
```diff
@@ -38,6 +38,14 @@ Refer to the documentation of the driver for the database you are using for the
 - MS SQL Server: `github.com/microsoft/go-mssqldb <https://github.com/microsoft/go-mssqldb>`_ (e.g. ``sqlserver://user:password@localhost:1433?database=dbname``)
 - SQLite (e.g. ``sqlite:file:/some/path/sqlite.db?_pragma=foreign_keys(1)&journal_mode(WAL)``)
 
+User sessions
+*************
+
+By default, user session data is stored in-memory, which does not survive restarts. You can configure persistent user session storage through ``storage.session.type``:
+
+- ``redis``: stores user session data in Redis, which is configured through ``storage.session.redis``. Recommended for high-volume deployments.
+- ``sql``: stores user session in the configured SQL database, which is configured through ``storage.sql``. Alternative option when Redis isn't available, but is less performant than Redis.
```
[the remark this answers is not in the document]
Comment (reinkrul) [reply docs/pages/deployment/storage.rst:47]: If using SQLite? In that case in-memory should probably be preferred?

[the remark this answers is not in the document]
Comment (reinkrul) [reply docs/pages/deployment/storage.rst:47]: If you want in-memory storage, you could now use sqlite in-memory. I very much doubt that it'll be noticable slower than the in-memory session store?

[the remark this answers is not in the document]
Comment (reinkrul) [reply docs/pages/deployment/storage.rst:47]: added documentation note on it

File: storage/session_sql.go (hunk 2 above)
Comment (reinkrul) [reply storage/session_sql.go:103]: I looked into using a database function for this, but there's no single approach (timestamp functions and/or datatypes) that works for all databases we support. Would it be better to just add a clock skew, e.g. 10 seconds (which should be plenty)?

**Pull request closed without merging**